charon/src/transform/index_intermediate_assigns.rs

```diff
@@ -0,0 +1,138 @@
+//! This micro-pass introduces intermediate assignments in preparation of
+//! [`index_to_function_calls`], so as to avoid borrow-checking errors.
+//!
+//! The problem comes from "and assignments" like in the snippet of code below:
+//! ```text
+//! x[0] += 1; // Desugars to: x[0] = copy x[0] + 1
+//! ```
+//!
+//! If we don't introduce an intermediate assignment, then the micro-pass which
+//! transforms the index operations into function calls generates the following
+//! LLBC:
+//! ```text
+//! dest = &mut x[0];
+//! src = & x[0]; // Invalidates dest
+//! *dest = copy (*src) + 1; // Can't dereference dest!
+//! ```
+//! (also note that the problem remains if we introduce `src` *then* `dest`).
+//!
+//! Our solution is to introduce a temporary variable for the assignment.
+//! We first transform the code to:
+//! ```text
+//! tmp = copy x[0] + 1;
+//! x[0] = move tmp;
+//! ```
+//!
+//! Which then allows us to transform it to:
+//! ```text
+//! // RHS:
+//! src = & x[0];
+//! tmp = copy (*src) + 1;
+//! // LHS:
+//! dest = &mut x[0];
+//! *dest = move tmp;
+//! ```
+
+use crate::ids::Vector;
+use crate::llbc_ast::*;
+use crate::transform::TransformCtx;
+use derive_visitor::{Drive, Visitor};
+
+use super::ctx::LlbcPass;
+
+impl Place {
+    fn contains_index(&self) -> bool {
+        for pe in &self.projection {
+            use ProjectionElem::*;
+            if let Index { .. } | Subslice { .. } = pe {
+                return true;
+            };
+        }
+        return false;
+    }
+}
+
+#[derive(Visitor)]
+#[visitor(Place(enter))]
+struct RvalueVisitor {
+    contains_index: bool,
+}
+
+impl RvalueVisitor {
+    fn enter_place(&mut self, p: &Place) {
+        self.contains_index = self.contains_index || p.contains_index();
+    }
+}
+
+impl Rvalue {
+    fn contains_index(&self) -> bool {
+        let mut v = RvalueVisitor {
+            contains_index: false,
+        };
+        self.drive(&mut v);
```

Comment:
if you wish, you can use `self.drive(&mut visitor_enter_fn(|p: &Place| contains_index |= p.contains_index()))` to avoid the need for a separate visitor struct